pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +279,176 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+			}
+			topoKeys.Insert(key)
+		}
+
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriverInfoStatuses tests that the specified CSIDriverInfoStatuses
+// have valid data.
+func validateCSIDriverInfoStatuses(
+	drivers []storage.CSIDriverInfoStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	driverNamesInStatuses := make(sets.String)
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoStatusVolumePluginMechanism(string(*driver.VolumePluginMechanism), idxPath.Child("volumePluginMechanism"))...)
+
+		// check for duplicate entries for the same driver in statuses
+		if driverNamesInStatuses.Has(driver.Name) {
+			idxPath := fldPath.Index(i)
+			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), driver.Name))
+		}
+		driverNamesInStatuses.Insert(driver.Name)
+	}
+
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecNodeID tests if Name in CSIDriverInfoSpec is a valid node id.
+func validateCSIDriverInfoSpecNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecTopologyKeys tests if TopologyKeys in CSIDriverInfoSpec are
+// valid topology keys.
+func validateCSIDriverInfoSpecTopologyKeys(keys []string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, key := range keys {
+		if len(key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, key))
+		}
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoStatusVolumePluginMechanism tests if VolumePluginMechanism in
+// CSIDriverInfoStatus is valid
+func validateCSIDriverInfoStatusVolumePluginMechanism(volumePluginMechanism string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !supportedVolumePluginMechanisms.Has(volumePluginMechanism) {
+		allErrs = append(allErrs, field.NotSupported(fldPath, volumePluginMechanism, supportedVolumePluginMechanisms.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriver validates a CSIDriver.
+func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(csiDriver.Name, field.NewPath("name"))...)
+
+	allErrs = append(allErrs, validateCSIDriverSpec(&csiDriver.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverUpdate validates a CSIDriver.
+func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
+	allErrs := ValidateCSIDriver(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
```

Comment:
@jsafrane do you expect podinfoonmount version to be mutable on a driver upgrade? Or do we expect the cluster-driver-registrar to delete and recreate the objects?